### PR TITLE
Feature/1364 contact key figures

### PR DESCRIPTION
**Pull request by timohuber**

### CODE REVIEW

**Comment by timohuber** [comment]
@mabiede Can you think of any other cases the tests do not cover?

File: pool/app/contact_counter/contact_counter.ml [hunk 1]
```diff
@@ -0,0 +1,85 @@
+let update_on_invitation_sent m =
+  Contact.
```
Comment (mabiede) [inline pool/app/contact_counter/contact_counter.ml:2]: Same as `Entity.increment_num_invitations`

File: pool/app/contact_counter/contact_counter.ml [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    { m with
+      num_invitations = m.num_invitations |> NumberOfInvitations.increment
+    }
+;;
+
+let update_on_session_signup m sessions =
```
Comment (mabiede) [inline pool/app/contact_counter/contact_counter.ml:8]: Same as `Entity.increment_num_assignments`

File: pool/app/contact_counter/contact_counter.ml [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  Contact.
+    { m with
+      num_assignments =
+        NumberOfAssignments.increment m.num_assignments (CCList.length sessions)
+    }
+;;
+
+let update_on_assignment_from_waiting_list = update_on_session_signup
```
Comment (mabiede) [inline pool/app/contact_counter/contact_counter.ml:16]: Use `Entity.XX` as well

File: pool/app/contact_counter/contact_counter.ml [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+  participated
+  increment_num_participaton
+  =
+  let open Contact in
+  let open Assignment in
+  match NoShow.value no_show, Participated.value participated with
+  | true, true ->
+    Error Pool_common.Message.(MutuallyExclusive Field.(Participated, NoShow))
+  | true, false -> Ok (increment_num_no_shows contact)
+  | false, true | false, false ->
```
Comment (mabiede) [inline pool/app/contact_counter/contact_counter.ml:30]:
```suggestion
  | false, (false | true) ->
```

File: pool/app/contact_counter/contact_counter.ml [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+     then increment_num_participations contact
+     else contact)
+    |> CCResult.return
+;;
+
+let update_on_session_cancellation assignments m =
+  Contact.
+    { m with
+      num_assignments =
+        NumberOfAssignments.decrement
```
Comment (mabiede) [inline pool/app/contact_counter/contact_counter.ml:44]: `Entity.decrement_num_assignments`

File: pool/app/contact_counter/contact_counter.ml [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+          m.num_assignments
+          (CCList.length assignments)
+    }
+;;
+
+let update_on_assignment_cancellation = update_on_session_cancellation
```
Comment (mabiede) [inline pool/app/contact_counter/contact_counter.ml:50]: same as above

File: pool/app/contact_counter/contact_counter.ml [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+      let closed_assignment =
+        closed_assignment
+        || CCOption.is_some no_show
+        || CCOption.is_some participated
+      in
+      let contact =
+        match no_show with
+        | None -> contact
+        | Some no_show ->
+          (match NoShow.value no_show with
```
Comment (mabiede) [inline pool/app/contact_counter/contact_counter.ml:71]:
```suggestion
          (if NoShow.value no_show then decrement_num_no_shows else decrement_num_show_ups) contact
```

File: pool/cqrs_command/invitation_command.ml [hunk 8]
```diff
@@ -53,10 +53,12 @@ end = struct
           ([ Invitation.Created (contacts, experiment) |> Pool_event.invitation
            ; Email.BulkSent emails |> Pool_event.email
            ]
-           @ CCList.map
-               (fun contact ->
-                 Contact.NumInvitationsIncreased contact |> Pool_event.contact)
-               contacts)
+           @ (contacts
+              |> CCList.map
+                   CCFun.(
```
Comment (mabiede) [inline pool/cqrs_command/invitation_command.ml:58]: may open `CCFun` at the beginning of the file and `CCList` within the function

File: pool/cqrs_command/session_command.ml [hunk 9]
```diff
@@ -421,8 +425,25 @@ end = struct
     let* (_ : unit list) =
       sessions |> CCList.map Session.is_cancellable |> CCList.all_ok
     in
+    let* (_ : unit list) =
+      let open CCList in
+      assignments
+      >|= (fun (_, lst) -> lst >|= Assignment.is_not_closed)
```
Comment (mabiede) [inline pool/cqrs_command/session_command.ml:431]:
```suggestion
      >|= snd
      >|= Assignment.is_not_closed
```

File: pool/cqrs_command/session_command.ml [hunk 10]
```diff
@@ -421,8 +425,25 @@ end = struct
...
+      |> flatten
+      |> all_ok
+    in
     let* emails =
-      contacts |> CCList.map (messages_fn command.reason) |> CCResult.flatten_l
+      assignments
+      |> CCList.map (fun (contact, _) -> contact |> messages_fn command.reason)
```
Comment (mabiede) [inline pool/cqrs_command/session_command.ml:437]:
```suggestion
      |> CCList.map (fst %> messages_fn command.reason)
```

File: pool/test/assignment_test.ml [hunk 11]
```diff
@@ -36,6 +36,27 @@ let confirmation_email contact =
     }
 ;;
 
+let update_contact_event
+  ?increment_assignments_by
```
Comment (mabiede) [inline pool/test/assignment_test.ml:40]: Add a default value for increment and decrement (e.g. 1)
```
  (* argument *)
  ?(increment_assignments_by = 1)

  (* map_or of fold_left *)
  CCOption.map_or ~default:num_assignments (fnc num_assignments)
```

KISS: refactor the function, that it takes one `int` as value with default `0` and uses increment when `> 0` and decrement when `< 0`.

File: pool/test/matcher_test.ml [hunk 12]
```diff
@@ -78,9 +83,14 @@ let create_invitations_repo _ () =
                ; Email.BulkSent emails |> Pool_event.email
                ]
                @ CCList.map
-                   (fun contact ->
-                     Contact.NumInvitationsIncreased contact
-                     |> Pool_event.contact)
+                   Contact.(
+                     fun ({ num_invitations; _ } as contact) ->
+                       { contact with
+                         num_invitations =
```
Comment (mabiede) [inline pool/test/matcher_test.ml:89]: use the entity function

File: pool/test/matcher_test.ml [hunk 13]
```diff
@@ -11,8 +11,13 @@ let expected_events experiment contacts create_message =
      ; Email.BulkSent emails |> Pool_event.email
      ]
      @ CCList.map
-         (fun contact ->
-           Contact.NumInvitationsIncreased contact |> Pool_event.contact)
+         Contact.(
+           fun ({ num_invitations; _ } as contact) ->
+             { contact with
+               num_invitations = NumberOfInvitations.increment num_invitations
```
Comment (mabiede) [inline pool/test/matcher_test.ml:17]: use the entity function

File: pool/test/invitation_test.ml [hunk 14]
```diff
@@ -29,10 +29,19 @@ let create () =
   in
   let expected =
     let email = Matcher_test.create_message contact |> CCResult.get_exn in
+    let contact_update =
+      Contact.
+        { contact with
+          num_invitations =
```
Comment (mabiede) [inline pool/test/invitation_test.ml:35]: use the entity function

File: pool/test/contact_counter_test.ml [hunk 15]
```diff
@@ -0,0 +1,441 @@
...
+let get_contact contact_id =
+  contact_id |> Contact.find database_label |> Lwt.map get_exn
+;;
+
+let get_session session_id =
+  session_id |> Session.find database_label |> Lwt.map get_exn
+;;
+
+let find_assignment_by_contact_and_session contact_id session_id =
+  let open Utils.Lwt_result.Infix in
```
Comment (mabiede) [inline pool/test/contact_counter_test.ml:16]: may open at the very top instead

**Review by mabiede — CHANGES_REQUESTED** [review]

File: pool/app/contact_counter/contact_counter.ml (hunk 2 above)
Comment (timohuber) [reply pool/app/contact_counter/contact_counter.ml:8]: It takes an argument, by hot many to increment

File: pool/app/contact_counter/contact_counter.ml (hunk 4 above)
Comment (timohuber) [reply pool/app/contact_counter/contact_counter.ml:30]: Nice 😃

File: pool/app/contact_counter/contact_counter.ml (hunk 5 above)
Comment (timohuber) [reply pool/app/contact_counter/contact_counter.ml:44]: Same as above, it takes an argument

File: pool/app/contact_counter/contact_counter.ml (hunk 6 above)
Comment (timohuber) [reply pool/app/contact_counter/contact_counter.ml:50]: As above, too 😄

File: pool/app/contact/entity.ml [hunk 16]
```diff
@@ -14,7 +14,7 @@ module NumberOfInvitations = struct
   let value m = m
   let of_int m = m
   let increment m = m + 1
-  let decrement m = if m > 0 then m - 1 else 0
+  let update step m = if m + step >= 0 then m + step else 0
```
Comment (mabiede) [inline pool/app/contact/entity.ml:17]:
```suggestion
  let update step m = m + step |> max 0
```

**Review by mabiede — APPROVED** [review]

**Pull request merged**